component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java

```diff
@@ -347,24 +349,41 @@ public static void setAuthorizedUser(OAuthTokenReqMessageContext tokenReqMsgCtx,
         AuthenticatedUser authenticatedUser = null;
 
         FederatedAssociationConfig federatedAssociationConfig = identityProvider.getFederatedAssociationConfig();
-        ClaimConfig claimConfig = identityProvider.getClaimConfig();
-
         ServiceProvider serviceProvider = getServiceProvider(tokenReqMsgCtx);
-        if (serviceProvider.getClaimConfig() != null &&
-                serviceProvider.getClaimConfig().isAlwaysSendMappedLocalSubjectId() &&
-                !Constants.LOCAL_IDP_NAME.equals(identityProvider.getIdentityProviderName())) {
-
-            if (federatedAssociationConfig.isEnabled()) {
-                String subjectIdentifier = resolveSubjectIdentifier(claimsSet,
-                        federatedAssociationConfig.getLookupAttributes());
-                User localUser = getLocalUser(tokenReqMsgCtx, subjectIdentifier);
-                if (localUser != null) {
-                    if (!isUserAssociated(localUser, identityProvider, subjectIdentifier)) {
-                        createAssociation(localUser, identityProvider, subjectIdentifier);
-                    }
-                    authenticatedUser = new AuthenticatedUser(localUser);
+        ClaimConfig serviceProviderClaimConfig = serviceProvider.getClaimConfig();
+        if (!Constants.LOCAL_IDP_NAME.equals(identityProvider.getIdentityProviderName())) {
+            SendLocalUser sendLocalUser = getSendLocalUser(serviceProviderClaimConfig);
+            User localUser = null;
+            if (federatedAssociationConfig != null && federatedAssociationConfig.isEnabled()) {
+                Map<String, String> mappedLocalClaims = resolveMappedLocalClaims(claimsSet,
+                        federatedAssociationConfig.getLookupAttributes(),
+                        identityProvider,
+                        tokenReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain());
+
+                localUser = getLocalUser(tokenReqMsgCtx, mappedLocalClaims);
+                if (localUser != null &&
+                        !isUserAssociated(localUser, identityProvider, authenticatedSubjectIdentifier)) {
+                    createAssociation(localUser, identityProvider, authenticatedSubjectIdentifier);
                 }
             }
+
+            switch (sendLocalUser) {
+                case OPTIONAL:
+                    if (localUser != null) {
+                        authenticatedUser = new AuthenticatedUser(localUser);
+                    }
+                    break;
+                case MANDATORY:
+                    if (localUser != null) {
+                        authenticatedUser = new AuthenticatedUser(localUser);
+                    } else {
+                        throw new IdentityOAuth2Exception(OAuth2ErrorCodes.INVALID_CLIENT,
```

Comment:
Is this the correct error code?

Reply:
Error response are handled in identity-inbound-auth-oauth [1]. To send a 401 response to the client we should set this error code

[1] - https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java#L288